gh-112606: Use `pthread_cond_timedwait_relative_np` in parking_lot.c when available

mattprodani opened the pull request:
Updates parking_lot.c to prefer functions that support CLOCK_MONOTONIC for semaphore and pthread_cond_t based implementations of _PySemaphore_PlatformWait. 


_Updated_: to instead use  [`pthread_cond_timedwait_relative_np`](https://github.com/apple-oss-distributions/libpthread/blob/d8c4e3c212553d3e0f5d76bb7d45a8acd61302dc/include/pthread/pthread.h#L556-L559) with a relative timeout which is supported on MacOS, unlike `pthread_condattr_setclock(..., CLOCK_MONOTONIC)`. Part of work to make `parking_lot.c` timing more accurate.

Linked to [PR](https://github.com/python/cpython/pull/112733) for POSIX semaphores.
<!-- gh-issue-number: gh-112606 -->
* Issue: gh-112606
<!-- /gh-issue-number -->

colesbury commented on Python/parking_lot.c:
```diff
@@ -67,10 +82,17 @@ _PySemaphore_Init(_PySemaphore *sema)
 #else
+    #if defined(CONDATTR_MONOTONIC)
+        if (!condattr_ptr){
+            pthread_condattr_init(&condattr);
```
We should probably use the same fields in `_PyRuntime.threads._condattr_monotonic` as `thread_pthread.h`. That will fix the failing `check-c-globals.py` test.

We also need to do the initialization in a thread-safe manner, which is also a bit tricky.

mattprodani replied:
So we could add a similar runtime object to the one `thread_pthread.h` uses. Might also be worth looking at `pthread_cond_clockwait` which allows passing the clock ID when the call is made rather than relying on the initializing the condition variable's attributes. It's was introduced with `sem_clockwait` in *glibc-2.30* and would work similarly to the semaphore implementation.

colesbury replied:
> So we could add a similar runtime object to the one thread_pthread.h

I'd like to use the exact same fields as the ones used by `thread_pthread.h` and just refactor the initialization code to make it available to `Python/parking_lot.c`.

Something like:

In `pycore_pythread.h`:

```
// Get and maybe initialize _condattr_monotonic
static inline pthread_condattr_t _PyThread_get_condattr_monotonic(void) {...};
// Initializes _condattr_monotonic (implemented in `thread_pthread.h`)
extern pthread_condattr_t _PyThread_init_condattr_monotonic(void);
```

With the appropriate build guards (`_USE_PTHREADS` and `CONDATTR_MONOTONIC`) as needed.

> Might also be worth looking at pthread_cond_clockwait...

I'd like to stick with the APIs we're already using in `thread_pthread.h` to avoid any new platform specific issues. Also, when compiling with glibc, `parking_lot.c` only uses the semaphore code path (not pthreads.)

colesbury replied:
On second thought, I don't think we should bother with `pthread_condattr_setclock` in `parking_lot.c`. The pthreads code path here is only used by macOS, and macOS doesn't support that call. Instead let's use [`pthread_cond_timedwait_relative_np`](https://github.com/apple-oss-distributions/libpthread/blob/d8c4e3c212553d3e0f5d76bb7d45a8acd61302dc/include/pthread/pthread.h#L556-L559) if available, which takes in a relative timeout and so avoids the clock issues. We'll need a `configure` check to check for it, but otherwise it should be more straightforward and effective.

mattprodani replied:
I'll get on that instead, explains why I couldn't figure out the point of `setclock` in testing.

colesbury commented on Python/parking_lot.c:
```diff
@@ -67,10 +82,17 @@ _PySemaphore_Init(_PySemaphore *sema)
+    #if defined(CONDATTR_MONOTONIC)
+        if (!condattr_ptr){
+            pthread_condattr_init(&condattr);
+            pthread_condattr_setclock(&condattr, CLOCK_MONOTONIC);
```
I think you also need to check the error code of `pthread_condattr_setclock` like `thread_pthread.h` does. Possibly there are platforms that have `pthread_condattr_setclock` implemented but don't support it with `CLOCK_MONOTONIC`?

mattprodani replied:
POSIX says `pthread_condattr_setclock` fails if the clock is unknown. This should be taken care of by the check for `CLOCK_MONOTONIC` but we can still implement a similar check for the return value. I think it would make sense since `CLOCK_MONOTONIC` can be supported at compile time but not runtime.

```
The pthread_condattr_setclock() function may fail if:

[EINVAL]
The value specified by clock_id does not refer to a known clock, or is a CPU-time clock.
These functions shall not return an error code of [EINTR].
```

colesbury reviewed (commented):
The semaphore bits look good. The `pthread_cond_timedwait` parts need some work and might be a bit tricky. If you like, you can split it into two PRs, or keep as one if you prefer.

mattprodani committed:
Use `pthread_cond_timedwait_relative_np` for MacOS futex implementation in
`parking_lot.c`

Adds a configure define for `HAVE_PTHREAD_COND_TIMEDWAIT_RELATIVE_NP` and
replaces `pthread_cond_timedwait` with `pthread_cond_timedwait_relative_np`
for relative time when supported in semaphore waiting logic.

colesbury commented on Python/parking_lot.c:
```diff
@@ -155,14 +155,19 @@ _PySemaphore_PlatformWait(_PySemaphore *sema, _PyTime_t timeout)
 #else
     pthread_mutex_lock(&sema->mutex);
     int err = 0;
+
```
Please remove the unrelated newline change, but otherwise looks good.

mattprodani committed:
remove newline

erlend-aasland reviewed (approved):
Configure changes LGTM. I'll leave parking lot for Sam :)

colesbury reviewed (approved):
This LGTM too.

As Victor pointed out, there similar logic in `Python/thread_pthread.h`, but I think that refactoring out the common logic is a bit complicated and better left to a subsequent change.
